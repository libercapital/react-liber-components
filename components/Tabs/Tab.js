'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tab = function Tab(_ref) {
  var label = _ref.label,
      children = _ref.children;
  return _react2.default.createElement(
    'div',
    { 'data-ref': label },
    children
  );
};

Tab.propTypes = {
  label: _propTypes2.default.string,
  children: _propTypes2.default.oneOfType([_propTypes2.default.any])
};

Tab.defaultProps = {
  label: '',
  children: ''
};

exports.default = Tab;