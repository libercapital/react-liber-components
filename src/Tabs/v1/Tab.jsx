import React from 'react';
import PropTypes from 'prop-types';

const Tab = ({ label, children }) => <div data-ref={label}>{children}</div>;

Tab.propTypes = {
  label: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.any]),
};

Tab.defaultProps = {
  label: '',
  children: '',
};

export default Tab;
