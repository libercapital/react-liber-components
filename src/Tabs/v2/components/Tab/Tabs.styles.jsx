import styled, { css } from 'styled-components';
import { handleThemeFromObject as getTheme } from '../../../../Themes';

export const TabContainer = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  /* overflow: hidden; */
  overflow: ${({transitioning}) => transitioning ? 'hidden' : null};
`;

const handleActive = (active = 0) => `-${active * 100}%`;

export const TabContentDisplay = styled.div`
  display: flex;
  transition: margin-left 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms;
  will-change: margin-left;
  margin-left: ${props => handleActive(props.active)};
  flex-direction: row;
  width: 100%;
`;

const hide = css`
  height: 0;
  padding: 0 !important;
  overflow: hidden;
  opacity: 0;
  pointer-events: none;
`;

export const TabContent = styled.div`
  flex-shrink: 0;
  width: 100%;
  transition: opacity 0.45s;
  ${props => !props.active && hide};
  color: ${props => getTheme(props.theme, 'colors.dark-text70', '#798e9b')};
`;

export const HeaderContainer = styled.div`
  position: relative;
`;

export const Slider = styled.span.attrs(({ active = 0, size = 225 }) => {
  const position = 24 + active * size;
  return {
    style: {
      '--slider-position': `${position}px`,
      '--slider-size': `${size}px`,
    },
  };
})`
  position: absolute;
  width: var(--slider-size);
  height: 0px;
  border-bottom: 2px solid
    ${props => getTheme(props.theme, 'colors.primary', '#009dff')};
  left: var(--slider-position);
  bottom: 0px;
  transition: left 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms,
    width 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms;
  will-change: left, width;
`;
