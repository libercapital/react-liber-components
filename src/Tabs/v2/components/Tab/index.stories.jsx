import React from 'react';
import styled from 'styled-components';
import { Tabs, Tab, DropDownTab } from '../../../tabsComponents';

const Container = styled.div`
  margin: 30px;
`;

const Content = styled.div`
  height: 320px;
  width: 100%;
`;

class TabsComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      activeTab: 0,
    };
  }

  handleChangeTab = index => {
    this.setState({ activeTab: index });
  };

  render() {
    return (
      <Tabs
        type="center"
        version="2"
        onChange={this.handleChangeTab}
        activeTab={this.state.activeTab}
      >
        <Tab version="2" label="Todos">
          Todos Content
        </Tab>
        <Tab version="2" label="Disponível" tooltip="tooltip">
          Disponível Content
        </Tab>
        <Tab version="2" label="Em Negociação">
          Em Negociação Content
        </Tab>
        <Tab version="2" label="Antecipados">
          Antecipados Content
        </Tab>
        <Tab version="2" label="Vencidos">
          Vencidos Content
        </Tab>
      </Tabs>
    );
  }
}

export default {
  title: 'Components 2.0|Tabs',
  parameters: {
    jest: ['Tabs'],
  },
};

export const center = () => (
  <Container>
    <Tabs version="2">
      <Tab version="2" label="Todos">
        Todos Content
      </Tab>
      <Tab version="2" label="Disponível" tooltip="tooltip">
        Disponível Content
      </Tab>
      <Tab version="2" label="Em Negociação">
        Em Negociação Content
      </Tab>
      <Tab version="2" label="Antecipados">
        Antecipados Content
      </Tab>
      <Tab version="2" label="Vencidos">
        Vencidos Content
      </Tab>
    </Tabs>
  </Container>
);

export const withStrongBorder = () => {
  return (
    <Container>
      <Tabs strongBorder version="2">
        <Tab version="2" label="Todos">
          Todos Content
        </Tab>
        <Tab version="2" label="Disponível" tooltip="tooltip">
          Disponível Content
        </Tab>
        <Tab version="2" label="Em Negociação">
          Em Negociação Content
        </Tab>
        <Tab version="2" label="Antecipados">
          Antecipados Content
        </Tab>
        <Tab version="2" label="Vencidos">
          Vencidos Content
        </Tab>
      </Tabs>
    </Container>
  );
};

export const controlledTabs = () => (
  <Container>
    <TabsComponent />
  </Container>
);

export const left = () => (
  <Container>
    <Tabs version="2" type="left">
      <Tab version="2" label="Todos">
        Todos Content
      </Tab>
      <Tab version="2" label="Disponível" tooltip="tooltip">
        Disponível Content
      </Tab>
      <Tab version="2" label="Em Negociação">
        Em Negociação Content
      </Tab>
      <Tab version="2" label="Antecipados">
        Antecipados Content
      </Tab>
      <Tab version="2" label="Vencidos">
        Vencidos Content
      </Tab>
      <Tab version="2" label="Vencidos">
        Vencidos Content
      </Tab>
    </Tabs>
  </Container>
);

export const withDropDown = () => {
  return (
    <Container>
      <Tabs version="2" strongBorder>
        <Tab version="2" label="Disponível" tooltip="tooltip">
          <Content>Disponível Content</Content>
        </Tab>
        <Tab version="2" label="Em Negociação">
          <Content>Em Negociação Content</Content>
        </Tab>
        <Tab version="2" label="Antecipados">
          <Content>Antecipados Content</Content>
        </Tab>
        <DropDownTab minWidth={275} strongBorder label="Aguardando">
          <Tab version="2" label="Aguardando Aceite">
            <Content>Aguardando Aceite Content</Content>
          </Tab>
          <Tab version="2" label="Aguardando Formalização">
            <Content>Aguardando Formalização Content</Content>
          </Tab>
          <Tab version="2" label="Aguardando Pagamento">
            <Content>Aguardando Pagamento Content</Content>
          </Tab>
        </DropDownTab>
        <Tab version="2" label="Vencidos">
          <Content>Vencidos Content</Content>
        </Tab>
      </Tabs>
    </Container>
  );
};
