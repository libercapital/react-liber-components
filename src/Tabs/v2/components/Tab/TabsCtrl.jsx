import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { TabContainer, TabContentDisplay, TabContent } from './Tabs.styles';

const isBetween = (value, min, max) => value >= min && value <= max;

const TabsCtrl = ({ tabs, active: ctrlActive, header }) => {
  const [transitioning, setTransitioning] = React.useState(false);
  const [isFirstLoad, setIsFirstLoad] = useState(true);

  const active = isBetween(ctrlActive, 0, tabs.length - 1) ? ctrlActive : 0;
  const mappedChildren = tabs.map((tab, index) => (
    <TabContent key={tab.key} active={active === index}>
      {tab.content}
    </TabContent>
  ));

  React.useEffect(() => {
    if (!isFirstLoad) {
      setTransitioning(true);
    } else {
      setIsFirstLoad(false);
    }
  }, [ctrlActive]);

  return (
    <TabContainer transitioning={transitioning}>
      {header}
      <TabContentDisplay onTransitionEnd={() => setTransitioning(false)} active={active}>
        {mappedChildren}
      </TabContentDisplay>
    </TabContainer>
  );
};

TabsCtrl.propTypes = {
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      content: PropTypes.any,
    }),
  ).isRequired,
  active: PropTypes.number.isRequired,
  header: PropTypes.node.isRequired,
};

export default TabsCtrl;
