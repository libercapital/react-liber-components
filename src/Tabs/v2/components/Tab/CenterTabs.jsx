import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { uniqueId } from 'lodash';
import TabsCtrl from './TabsCtrl';
import { SimpleTabs } from '../SimpleTab';
import { metaProperty } from '@babel/types';

const handleClick = (index, setActive) => () => {
  setActive(index);
};

const setAction = (child, index, active, setActive, styleProps) => {
  const { label, children } = child.props;
  return {
    actionTab: [
      React.cloneElement(child, {
        key: uniqueId(`tab_${label}`),
        as: 'div',
        onClick: handleClick(index, setActive),
        active: active === index,
        children: label,
        ...styleProps,
      }),
    ],
    tab: [{ key: uniqueId(label), content: children }],
    counter: 1,
  };
};

const mapDropDownChildren = (active, setActive) => (current, child) => {
  const { actionTab, tab, counter } = setAction(child, current.index, active, setActive);

  return {
    mappedTabs: [...current.mappedTabs, ...actionTab],
    tabs: [...current.tabs, ...tab],
    index: current.index + counter,
  };
};

const setDropDownAction = (dropdown, index, active, setActive) => {
  const children = React.Children.toArray(dropdown.props.children);
  const { mappedTabs, tabs: tab, index: counter } = children.reduce(
    mapDropDownChildren(active, setActive),
    { mappedTabs: [], tabs: [], index },
  );
  return {
    actionTab: [
      React.cloneElement(dropdown, {
        children: mappedTabs,
        active: active >= index && active < counter,
        as: 'div',
      }),
    ],
    tab,
    counter: counter - index,
  };
};

const isDropDown = child => child.type.displayName === 'DropDownTab';

const mapChildren = (active, setActive, styleProps) => (current, child) => {
  const action = isDropDown(child) ? setDropDownAction : setAction;
  const { actionTab, tab, counter } = action(child, current.index, active, setActive, styleProps);

  return {
    mappedTabs: [...current.mappedTabs, ...actionTab],
    tabs: [...current.tabs, ...tab],
    index: current.index + counter,
  };
};

const Tabs = props => {
  const [active, setActive] = useState(0);
  const {
    children,
    activeTab = active,
    onChange = setActive,
    ControlComponent,
    ...styleProps
  } = props;

  const { mappedTabs, tabs } = React.Children.toArray(children).reduce(
    mapChildren(activeTab, onChange, styleProps),
    { mappedTabs: [], tabs: [], index: 0 },
  );

  const header = <SimpleTabs>{mappedTabs}</SimpleTabs>;
  return React.cloneElement(ControlComponent, {
    header: header,
    active: activeTab,
    tabs: tabs,
  });
};

Tabs.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]).isRequired,
  activeTab: PropTypes.number,
  onChange: PropTypes.func,
  ControlComponent: PropTypes.element,
};

Tabs.defaultProps = {
  activeTab: undefined,
  onChange: undefined,
  ControlComponent: <TabsCtrl header={<div />} active={0} tabs={[]} />,
};

export default Tabs;
