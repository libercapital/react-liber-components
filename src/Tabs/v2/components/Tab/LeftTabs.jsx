import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { uniqueId } from 'lodash';
import TabsCtrl from './TabsCtrl';
import { SimpleTabs } from '../SimpleTab';
import { HeaderContainer, Slider } from './Tabs.styles';

class Tabs extends Component {
  state = {
    active: 0,
    size: 0,
  };

  tabs = [];

  componentDidMount = () => {
    const { active } = this.state;
    const { activeTab = active } = this.props;

    setTimeout(() => {
      this.setState({ size: this.tabs[activeTab].offsetWidth });
    }, 0);
  };

  handleClick = active => () => {
    const { onChange } = this.props;
    const currentTab = this.tabs[active];
    onChange(active);
    this.setState({
      active,
      size: currentTab.offsetWidth,
    });
  };

  tabRef = index => element => {
    this.tabs[index] = element;
  };

  mapChildren = active => (current, child, index) => {
    const { label, children } = child.props;
    return {
      mappedTabs: [
        ...current.mappedTabs,
        React.cloneElement(child, {
          key: uniqueId(`tab_${label}`),
          ref: this.tabRef(index),
          left: true,
          showUnderline: false,
          as: 'button',
          onClick: this.handleClick(index),
          active: active === index,
          children: label,
        }),
      ],
      tabs: [...current.tabs, { key: uniqueId(label), content: children }],
    };
  };

  render() {
    const { active, size } = this.state;
    const { children, activeTab = active, ControlComponent } = this.props;
    const { mappedTabs, tabs } = React.Children.toArray(children).reduce(
      this.mapChildren(activeTab),
      { mappedTabs: [], tabs: [] },
    );
    const header = (
      <HeaderContainer>
        <SimpleTabs left>{mappedTabs}</SimpleTabs>
        <Slider active={activeTab} size={size} />
      </HeaderContainer>
    );
    return React.cloneElement(ControlComponent, {
      header: header,
      active: activeTab,
      tabs: tabs,
    });
  }
}

Tabs.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]).isRequired,
  activeTab: PropTypes.number,
  onChange: PropTypes.func,
  ControlComponent: PropTypes.element,
};

Tabs.defaultProps = {
  activeTab: undefined,
  onChange: () => {},
  ControlComponent: <TabsCtrl header={<div />} active={0} tabs={[]} />,
};

export default Tabs;
