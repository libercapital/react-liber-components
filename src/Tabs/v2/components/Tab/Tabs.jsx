import React from 'react';
import CenterTabs from './CenterTabs';
import LeftTabs from './LeftTabs';

const Tabs = ({ type = 'center', ...props }) =>
  ({
    center: <CenterTabs {...props} />,
    left: <LeftTabs {...props} />,
  }[type]);

export default Tabs;
