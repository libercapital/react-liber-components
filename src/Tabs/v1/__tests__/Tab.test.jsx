import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import Tab from '../Tab';

describe('Testing Tab Component', () => {
  let props;
  configure({ adapter: new Adapter() });

  beforeEach(() => {
    props = {
      label: 'test',
    };
  });
  it('should render correctly with element', () => {
    const wrapper = shallow(<Tab {...props}><div /></Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with string', () => {
    const wrapper = shallow(<Tab {...props}>Test</Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with number', () => {
    const wrapper = shallow(<Tab {...props}>{1212}</Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with arrayOf element', () => {
    const wrapper = shallow(<Tab {...props}><div /><div /></Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with arrayOf string', () => {
    const wrapper = shallow(<Tab {...props}>{['test1', 'test2']}</Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with arrayOf number', () => {
    const wrapper = shallow(<Tab {...props}>{['1212', '9898']}</Tab>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
