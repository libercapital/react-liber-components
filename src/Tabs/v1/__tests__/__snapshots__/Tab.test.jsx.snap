// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Tab Component should render correctly with arrayOf element 1`] = `
<div
  data-ref="test"
>
  <div />
  <div />
</div>
`;

exports[`Testing Tab Component should render correctly with arrayOf number 1`] = `
<div
  data-ref="test"
>
  1212
  9898
</div>
`;

exports[`Testing Tab Component should render correctly with arrayOf string 1`] = `
<div
  data-ref="test"
>
  test1
  test2
</div>
`;

exports[`Testing Tab Component should render correctly with element 1`] = `
<div
  data-ref="test"
>
  <div />
</div>
`;

exports[`Testing Tab Component should render correctly with number 1`] = `
<div
  data-ref="test"
>
  1212
</div>
`;

exports[`Testing Tab Component should render correctly with string 1`] = `
<div
  data-ref="test"
>
  Test
</div>
`;
